sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java

```diff
@@ -964,20 +964,23 @@ public Flux<OpenConnectionResponse> openConnectionsAndInitCaches(
                             .collect(Collectors.toList());
 
                     return Flux.fromIterable(addressInfos)
-                            .flatMap(addressInfo -> {
-                                this.serverPartitionAddressCache.set(addressInfo.getLeft(), addressInfo.getRight());
-
-                                if (this.openConnectionsHandler != null) {
-                                    return this.openConnectionsHandler.openConnections(
-                                            Arrays
-                                                .stream(addressInfo.getRight())
-                                                .map(addressInformation -> addressInformation.getPhysicalUri())
-                                                .collect(Collectors.toList()));
-                                }
-
-                                logger.info("OpenConnectionHandler is null, can not open connections");
-                                return Flux.empty();
-                            });
+                            .flatMap(
+                                addressInfo -> {
+                                    this.serverPartitionAddressCache.set(addressInfo.getLeft(), addressInfo.getRight());
+
+                                    if (this.openConnectionsHandler != null) {
+                                        return this.openConnectionsHandler.openConnections(
+                                                Arrays
+                                                    .stream(addressInfo.getRight())
+                                                    .map(addressInformation -> addressInformation.getPhysicalUri())
+                                                    .collect(Collectors.toList()));
+                                    }
+
+                                    logger.info("OpenConnectionHandler is null, can not open connections");
+                                    return Flux.empty();
+                                },
+                                Configs.getCPUCnt() * 10,
```

Comment:
hmm, since we have the semaphore control in the `openConnectionsHandler`, so i feel like it might be okie to not have concurrency control here. 

Is it to try to prevent the semaphore slim timeout?

Reply:
Goal was to not add too much concurrency on few CPUs - default is 256 with prefetch of 32. On 4 Cores that is probably too much already. Definitely on single core. But agree - makes sense to not artifically throttle. let me change the factor so that on 4 cores we reach at about 256. We can iterate on this if needed.

Reply:
Never mind - just aligned on CPU * 10 - the Semaphore was too restrictive. Opening multiple connections per core isn't a problem - it is mostly i/o bound. Keeping a limit is good - but more than 1 seems reasonable